proposals/0185-vote-account-v4.md

```diff
@@ -0,0 +1,245 @@
+
+Currently, all block revenue, including both transaction base fees and priority
+fees, is collected into a validator's node id account. This proposal details
+changes to the vote account state that will allow validators to specify how
+different sources of income are collected in future SIMD's. This proposal also
+updates the bookkeeping for authorized voters in the vote account state.
+
+### Vote Account 
+
+A new version of vote state will be introduced with the enum discriminant value
+of `3u32` which will be little endian encoded in the first 4 bytes of account
+data. 
+
+```rust
+pub enum VoteStateVersions {
+    V1(..),
+    V2(..),
+    V3(..),
+    V4(..), // <- new variant
+}
+```
+
+This new version of vote state will include new fields for setting the
+commission and collector account for the following sources of validator income:
+inflation rewards and block revenue. It will also remove the `prior_voters`
+field.
+
+```rust
+pub struct VoteStateV4 {
+    pub node_pubkey: Pubkey,
+    pub authorized_withdrawer: Pubkey,
+
+    /// REMOVED
+    /// commission: u8,
+
+    /// NEW: the collector accounts for validator income
+    pub inflation_rewards_collector: Pubkey,
+    pub block_revenue_collector: Pubkey,
+
+    /// NEW: basis points (0-10,000) that represent how much of each income
+    /// source should be given to this VoteAccount
+    pub inflation_rewards_commission_bps: u16,
+    pub block_revenue_commission_bps: u16,
+
+    /// NEW: reward amount pending distribution to stake delegators
+    pub pending_delegator_rewards: u64,
+
+    pub votes: VecDeque<LandedVote>,
+    pub root_slot: Option<Slot>,
+    pub authorized_voters: AuthorizedVoters,
+
+    /// REMOVED
+    /// prior_voters: CircBuf<(Pubkey, Epoch, Epoch)>,
+
+    pub epoch_credits: Vec<(Epoch, u64, u64)>,
+    pub last_timestamp: BlockTimestamp,
+}
+```
+
+### Vote Program
+
+All vote instructions MUST be updated to support deserializing v4 vote accounts.
+
+Whenever a vote account is initialized OR modified by the vote program in a
+transaction AND hasn't been updated to v4 yet, the account state MUST be saved
+in the new format with the following default values for the new fields described
+above:
+
+```rust
+VoteStateV4 {
+    // ..
+
+    inflation_rewards_collector: Pubkey::default(),
+    block_rewards_collector: Pubkey::default(),
+    inflation_rewards_commission_bps: 100u16 * (old_vote_state.commission as u16),
+    block_rewards_commission_bps: 10_000u16,
+    pending_delegator_rewards: 0u64,
+
+    // ..
+}
+```
+
+If a modified vote account's size is smaller than `3762` bytes, first resize the
+account to `3762` bytes before updating the account data. The vote program does
+not need to check if the resulting account is rent exempt, the runtime will
+enforce that check. This differs from the prior vote program implementation which
+falls back to store vote state as v2 if the account size cannot be resized while
+keeping the account rent exempt.
```

Comment:
I felt this was nice to have because it simplifies the vote program implementation and ensures that all v4 vote accounts have the same size. Here's the current code in question that this will clean up:

https://github.com/anza-xyz/agave/blame/0df4a2d963e06ced795247719ef1aa16c6d71d15/programs/vote/src/vote_state/mod.rs#L40-L54